ipa-core/src/protocol/basics/validate.rs

```diff
@@ -0,0 +1,218 @@
+use std::{
+    convert::Infallible,
+    marker::PhantomData,
+    pin::Pin,
+    task::{Context as TaskContext, Poll},
+};
+
+use futures::{
+    future::try_join,
+    stream::{Fuse, Stream, StreamExt},
+    Future, FutureExt,
+};
+use generic_array::GenericArray;
+use pin_project::pin_project;
+use sha2::{
+    digest::{typenum::Unsigned, FixedOutput, OutputSizeUser},
+    Digest, Sha256,
+};
+
+use crate::{
+    error::Error,
+    ff::Serializable,
+    helpers::{Direction, Message},
+    protocol::{context::Context, RecordId},
+    secret_sharing::{replicated::ReplicatedSecretSharing, SharedValue},
+};
+
+type HashFunction = Sha256;
+type HashSize = <HashFunction as OutputSizeUser>::OutputSize;
+type HashOutputArray = [u8; <HashSize as Unsigned>::USIZE];
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct HashValue(GenericArray<u8, HashSize>);
+
+impl Serializable for HashValue {
+    type Size = HashSize;
+    type DeserializationError = Infallible;
+
+    fn serialize(&self, buf: &mut GenericArray<u8, Self::Size>) {
+        buf.copy_from_slice(self.0.as_slice())
+    }
+
+    fn deserialize(buf: &GenericArray<u8, Self::Size>) -> Result<Self, Self::DeserializationError> {
+        Ok(Self(buf.to_owned()))
+    }
+}
+
+impl Message for HashValue {}
+
+struct ReplicatedValidatorFinalization<C> {
+    f: Pin<Box<dyn Future<Output = Result<(), Error>>>>,
+    ctx: C,
+}
+
+impl<C: Context + 'static> ReplicatedValidatorFinalization<C> {
+    fn new(active: ReplicatedValidatorActive<C>) -> Self {
+        let ReplicatedValidatorActive {
+            ctx,
+            left_hash,
+            right_hash,
+        } = active;
+        // Ugh: The version of sha2 we currently use doesn't use the same GenericArray version as we do.
+        let left_hash = HashValue(GenericArray::from(<HashOutputArray>::from(
+            left_hash.finalize_fixed(),
+        )));
+        let right_hash = HashValue(GenericArray::from(<HashOutputArray>::from(
+            right_hash.finalize_fixed(),
+        )));
+        let left_peer = ctx.role().peer(Direction::Left);
+        let right_peer = ctx.role().peer(Direction::Left);
+        let ctx_ref = &ctx;
```

Comment:
do you want to borrow here or it is acceptable to clone the context and move it inside the box?

If we want to borrow from `C`, we would need to create a self-referential struct, so maybe we need an inner struct that is pinned and holds `f` that borrows from `ctx` and `ctx`


however, I was able to compile it just by `let ctx = ctx.clone()` but that's probably not what you want here

Reply:
Yeah, I managed to get this to compile with a similar trick.  Though the tests are now failing because the `TestWorld` doesn't have a `'static` lifetime.  I put the `'static` condition on to avoid problems creating the future, ...